pkg/retention/retention.go

```diff
@@ -416,3 +419,107 @@ func DeleteMetricsSegmentData(mmetaFile string, metricSegmentsToDelete map[strin
+	err := syscall.Statfs(filepath.Clean(dataPath), &fsStats)
+	if err != nil {
+		log.Errorf("doInodeBasedDeletion: Failed to get inode stats: %v", err)
+		return
+	}
+
+	totalInodes := fsStats.Files
+	freeInodes := fsStats.Ffree
+	usedInodes := totalInodes - freeInodes
+
+	if totalInodes == 0 {
+		log.Errorf("doInodeBasedDeletion: Invalid total inodes count: %d", totalInodes)
+		return
+	}
+
+	currentUsagePercent := (usedInodes * 100) / totalInodes
+
+	log.Infof("doInodeBasedDeletion: Current inode usage: %v%% (%d used of %d total), Max allowed: %v%%, IngestNodeDir: %v",
+		currentUsagePercent, usedInodes, totalInodes, maxInodeUsagePercent, ingestNodeDir)
+
+	if currentUsagePercent <= maxInodeUsagePercent {
+		return
+	}
+
+	if deletionWarningCounter < MAXIMUM_WARNINGS_COUNT {
+		log.Warnf("Skipping inode-based deletion since try %d, Current usage: %v%%, Max allowed: %v%%",
+			deletionWarningCounter, currentUsagePercent, maxInodeUsagePercent)
+		return
+	}
+
+	// Get all segments sorted by time
+	allSegMetas := writer.ReadLocalSegmeta(false)
+	currentMetricsMeta := path.Join(ingestNodeDir, mmeta.MetricsMetaSuffix)
+	allMetricMetas, err := mmeta.ReadMetricsMeta(currentMetricsMeta)
+	if err != nil {
+		log.Errorf("doInodeBasedDeletion: Failed to get metric meta entries, filepath=%v, err: %v", currentMetricsMeta, err)
+		return
+	}
+
+	// Combine and sort all entries by time
+	allEntries := make([]interface{}, 0, len(allMetricMetas)+len(allSegMetas))
+	for i := range allMetricMetas {
+		allEntries = append(allEntries, allMetricMetas[i])
+	}
+	for i := range allSegMetas {
+		allEntries = append(allEntries, allSegMetas[i])
+	}
+
+	sort.Slice(allEntries, func(i, j int) bool {
+		var timeI uint64
+		if segMeta, ok := allEntries[i].(*structs.SegMeta); ok {
+			timeI = segMeta.LatestEpochMS
+		} else if metricMeta, ok := allEntries[i].(*structs.MetricsMeta); ok {
+			timeI = uint64(metricMeta.LatestEpochSec * 1000)
+		} else {
+			log.Errorf("doInodeBasedDeletion: Unexpected entry type in allEntries: %T", allEntries[i])
+			return false
+		}
+
+		var timeJ uint64
+		if segMeta, ok := allEntries[j].(*structs.SegMeta); ok {
+			timeJ = segMeta.LatestEpochMS
+		} else if metricMeta, ok := allEntries[j].(*structs.MetricsMeta); ok {
+			timeJ = uint64(metricMeta.LatestEpochSec * 1000)
+		} else {
+			log.Errorf("doInodeBasedDeletion: Unexpected entry type in allEntries: %T", allEntries[j])
+			return false
+		}
+		return timeI < timeJ
+	})
+
+	// Delete oldest segments until we get below the threshold
+	segmentsToDelete := make(map[string]*structs.SegMeta)
+	metricSegmentsToDelete := make(map[string]*structs.MetricsMeta)
+	inodesToFree := usedInodes - (totalInodes * maxInodeUsagePercent / 100)
+	segmentsMarked := uint64(0)
+
+	for _, metaEntry := range allEntries {
+		if segmentsMarked >= inodesToFree {
+			break
+		}
+		switch entry := metaEntry.(type) {
+		case *structs.MetricsMeta:
+			metricSegmentsToDelete[entry.MSegmentDir] = entry
+			segmentsMarked++
```

Comment:
Deleting a segment is currently being treated as removing a single `inode`, but that's not accurate. The `Delete` function actually deletes the entire segment directory, which may contain multiple inodes. It would be better to calculate the inode count for each segment directory and increment the required inode deletions by that amount when marking a segment as deleted.